File: Design-Patterns-Using-CSharp/Proxy/ProxyPatternLazy.cs
using System;

// BookParser Interface
interface IBookParser
{
    int NumPages();
}

// ConcreteBookParser Class
class ConcreteBookParser : IBookParser
{
    private int numPages;

    public ConcreteBookParser()
    {
        Console.WriteLine("Concrete Book Parser Created");
        // Number of pages calculation heavy operation.
        // Suppose this calculation results in 1000 pages.
        this.numPages = 1000;
    }

    public int NumPages()
    {
        Console.WriteLine("Concrete Book Parser Request Method");
        return this.numPages;
    }
}

// LazyBookParserProxy Class
class LazyBookParserProxy : IBookParser
{
    private ConcreteBookParser subject;

    public int NumPages()
    {
        if (subject == null)
        {
            subject = new ConcreteBookParser();
        }
        return subject.NumPages();
    }
}

// Client code
class ProxyPatternLazy
{
    static void Main(string[] args)
    {
        LazyBookParserProxy proxy = new LazyBookParserProxy();
        Console.WriteLine(proxy.NumPages());
    }
}

/*
Concrete Book Parser Created
Concrete Book Parser Request Method
1000
*/

File: Design-Patterns-Using-Java/Proxy/ProxyPatternLazy.java
interface BookParser {
    int numPages();
}

class ConcreteBookParser implements BookParser {
    private int numPages;

    public ConcreteBookParser() {
        System.out.println("Concrete Book Parser Created");
        // Number of pages calculation heavy operation.
        // Suppose this calculation results in 1000 pages.
        this.numPages = 1000;
    }

    @Override
    public int numPages() {
        System.out.println("Concrete Book Parser Request Method");
        return this.numPages;
    }
}

class LazyBookParserProxy implements BookParser {
    private ConcreteBookParser subject;

    @Override
    public int numPages() {
        if (subject == null) {
            subject = new ConcreteBookParser();
        }
        return subject.numPages();
    }
}

public class ProxyPatternLazy {
    public static void main(String[] args) {
        LazyBookParserProxy proxy = new LazyBookParserProxy();
        System.out.println(proxy.numPages());
    }
}

/* 
Concrete Book Parser Created
Concrete Book Parser Request Method
1000
*/
